File: mybatis/src/main/java/data/spring/mybatis/application/service/ProductService.java
package data.spring.mybatis.application.service;

import data.spring.mybatis.application.required.ProductRepository;
import data.spring.mybatis.application.required.ProductUseCase;
import data.spring.mybatis.application.service.command.ProductSearchCommand;
import data.spring.mybatis.application.service.command.ProductUpdateCommand;
import data.spring.mybatis.domain.Product;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

public class ProductService implements ProductUseCase {
    private final ProductRepository productRepository;

    public ProductService(ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    @Override
    public Optional<Product> findById(Long productId) {
        return this.productRepository.findById(productId);
    }

    @Override
    public List<Product> findAll(ProductSearchCommand searchCommand) {
        return this.productRepository.findAll(searchCommand);
    }

    @Override
    public void save(Product product) {
        this.productRepository.save(product);
    }

    @Transactional
    @Override
    public int saveAll(List<Product> entities) {
        return this.productRepository.saveAll(entities);
    }

    @Override
    public void update(ProductUpdateCommand updateCommand) {
        this.productRepository.update(updateCommand);
    }

    @Transactional
    @Override
    public int updateAll(List<ProductUpdateCommand> updateCommands) {
        int cnt = 0;
        for (ProductUpdateCommand command : updateCommands) {
            this.productRepository.update(command);
            cnt++;
        }
        return cnt;
    }
}

File: mybatis/src/main/java/data/spring/mybatis/application/service/product/ProductService.kt
package data.spring.mybatis.application.service.product

import data.spring.mybatis.application.provided.product.ProductUseCase
import data.spring.mybatis.application.required.product.ProductRepository
import data.spring.mybatis.application.service.product.command.ProductSearchCommand
import data.spring.mybatis.application.service.product.command.ProductUpdateCommand
import data.spring.mybatis.application.exception.NoDataFoundException
import data.spring.mybatis.domain.product.Product
import org.springframework.transaction.annotation.Transactional

open class ProductService(
    val productRepository: ProductRepository
): ProductUseCase {
    override fun findById(productId: Long): Product?
        = this.productRepository.findById(productId) ?: throw NoDataFoundException("Product not found with id: ${productId}.")

    override fun findAll(searchCommand: ProductSearchCommand): List<Product>
        = this.productRepository.findAll(searchCommand)
            .ifEmpty { throw NoDataFoundException("No products found for search command: ${searchCommand}.") }

    override fun save(product: Product)
        = this.productRepository.save(product)

    override fun saveAll(products: List<Product>): Int
        = this.productRepository.saveAll(products)

    override fun update(updateCommand: ProductUpdateCommand)
        = this.productRepository.update(updateCommand)

    @Transactional
    override fun updateAll(updateCommands: List<ProductUpdateCommand>): Int
        = updateCommands.onEach(this.productRepository::update).count()
}